File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int N;
    static int M;
    static Edge[] e;
    static int[] parent;

    static class Edge implements Comparable<Edge> {
        int from;
        int to;
        int cost;

        public Edge(int from, int to, int cost) {
            this.from = from;
            this.to = to;
            this.cost = cost;
        }

        @Override
        public int compareTo(Edge o) {
            return this.cost - o.cost;
        }
    }

    public static int find(int x){
        if(parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    public static boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if(rootX == rootY) return false;

        if(rootX < rootY){
            parent[rootX] = rootY;

        }else{
            parent[rootY] = rootX;
        }

        return true;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(bf.readLine());
        N = Integer.parseInt(st.nextToken());
        parent = new int[N+1];
        for (int i = 0; i < N+1; i++) {
            parent[i] = i;
        }

        st = new StringTokenizer(bf.readLine());
        M = Integer.parseInt(st.nextToken());
        e = new Edge[M];

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(bf.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());

            e[i] = new Edge(a, b, c);
        }

        Arrays.sort(e);

        int result = 0;

        for (int i = 0; i < M; i++) {
            if(union(e[i].from, e[i].to)) {
                // 연결 가능
                result += e[i].cost;
            }
        }

        System.out.println(result);
    }
}

File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.py
import sys

input = sys.stdin.readline

N = int(input())
M = int(input())

parent = [i for i in range(N+1)] # 각 노드의 부모

def find(x):
    if parent[x] == x:
        # 집합의 대표 노드 return
        return x
    else:
        parent[x] = find(parent[x])
        return parent[x]

def union(x, y):
    p1 = find(x)
    p2 = find(y)

    if p1 == p2:
        return

    # 둘의 부모가 다르면 합치기 가능
    if p1 < p2:
        parent[p2] = p1
    else:
        parent[p1] = p2

graph = [] # 간선을 비용순으로 정렬하면 되므로 2차원으로 만들면 됨
for _ in range(M):
    a, b, c = map(int, input().split())
    graph.append((a, b, c))

graph.sort(key = lambda x: x[2]) # 비용순 정렬

ans = 0
for a, b, c in graph:
    if find(a) != find(b):
        ans += c
        union(a, b)

print(ans)